Commit for #23 - Implement synthetic traffic filtering from GSM

gzepeda opened the pull request:
We are sending the changes to resolve issue #23

This was worked with a team composed by @georgejolema , @oscargarciacolon and myself.

abaranch commented on src/Microsoft.ApplicationInsights.AspNet/TelemetryInitializers/SyntheticTelemetryInitializer.cs:
```diff
@@ -0,0 +1,60 @@
+                }
+            } else
+            {
+                AspNetEventSource.Instance.SyntheticTelemetryInitializerOnInitializeTelemetryHeaderNotPresent();
```
So customers set SyntheticSource manually. How is it related to "HeaderNotPresent" and why would you want to warn about it?..

gzepeda replied:
Interesting point of view. Now that you mention it... it should not be necessary to do it. We will remove the logging piece.

abaranch commented on src/Microsoft.ApplicationInsights.AspNet/TelemetryInitializers/SyntheticTelemetryInitializer.cs:
```diff
@@ -0,0 +1,60 @@
+                    }
+                } else
+                {
+                    telemetry.Context.Operation.SyntheticSource = null;
```
Why would you want to null it and how did become not null?..

gzepeda replied:
Just to make sure null stays null... 

If under any scenario that is going to be changed (even externally) then we will remove that set.

abaranch commented on src/Microsoft.ApplicationInsights.AspNet/TelemetryInitializers/SyntheticTelemetryInitializer.cs:
```diff
@@ -0,0 +1,60 @@
+    /// </summary>
+    public class SyntheticTelemetryInitializer : TelemetryInitializerBase
+    {
+        private const string HeaderNameDefault = "GsmSyntheticTestRunId";
```
We do not check this header in WEB sdk, are you sure it is there?... https://github.com/Microsoft/ApplicationInsights-dotnet-server/blob/master/Src/Web/Web.Shared.Net/WebTestTelemetryInitializer.cs

gzepeda replied:
The issue #23 specified the need to check for that specific header. If it is present or not I believe it would depend on the originator of the request to be tracked. 

Is there a different header that needs to be checked?

abaranch replied:
Please refer to the latest implementation we have in Web.

gzepeda replied:
Will remove that validation then and just validate if `telemetry.Context.Operation.SyntheticSource` is null or not.

abaranch commented on src/Microsoft.ApplicationInsights.AspNet/TelemetryInitializers/SyntheticTelemetryInitializer.cs:
```diff
@@ -0,0 +1,60 @@
+                {
+                    telemetry.Context.Operation.SyntheticSource = SyntheticSourceHeaderValue;
+
+                    var locationHeader = platformContext.Request.Headers[SyntheticTestLocation];
```
Can we set values only when all headers are present like we do in Web? https://github.com/Microsoft/ApplicationInsights-dotnet-server/blob/master/Src/Web/Web.Shared.Net/WebTestTelemetryInitializer.cs#L36

gzepeda replied:
Will change in a bit.

abaranch commented on src/Microsoft.ApplicationInsights.AspNet/TelemetryInitializers/SyntheticTelemetryInitializer.cs:
```diff
@@ -0,0 +1,60 @@
+                    var locationHeader = platformContext.Request.Headers[SyntheticTestLocation];
+                    if (!string.IsNullOrEmpty(headerValue))
+                    {
+                        telemetry.Context.User.Id = locationHeader;
```
You need concatenation  here for sampling to work: https://github.com/Microsoft/ApplicationInsights-dotnet-server/blob/master/Src/Web/Web.Shared.Net/WebTestTelemetryInitializer.cs#L42

gzepeda replied:
Will change in a bit.

gzepeda commented:
Modifications based on the comments sent by @abaranch added on the latest commit.

abaranch commented on test/Microsoft.ApplicationInsights.AspNet.Tests/TelemetryInitializers/SyntheticTelemetryInitializerTests.cs:
```diff
@@ -0,0 +1,101 @@
+    using Xunit;
+    using Microsoft.AspNet.Http.Internal;
+
+    public class SyntheticTelemetryInitializerTests
```
Please add unit tests that verify that user.id and session.id are set correctly

gzepeda replied:
Will do.

gzepeda commented on src/Microsoft.ApplicationInsights.AspNet/Extensibility/Implementation/Tracing/AspNetEventSource.cs:
```diff
@@ -123,6 +123,16 @@ public void LogComponentVersionTelemetryInitializerFailsToAccessProjectJson(stri
+        /// </summary>
+        /// <param name="appDomainName">An ignored placeholder to make EventSource happy.</param>
+        [Event(9, Message = "SyntheticTelemetryInitializer.OnInitializeTelemetry - telemetry.Context.Operation.SyntheticSource is already set, returning.", Level = EventLevel.Warning, Keywords = Keywords.Diagnostics)]
+        public void SyntheticTelemetryInitializerOnInitializeTelemetryHeaderNotPresent(string appDomainName = "Incorrect")
```
Will remove.

gzepeda commented:
Latest changes done based on @abaranch input: 
- Based implementation on the Web Version [https://github.com/Microsoft/ApplicationInsights-dotnet-server/blob/master/Src/Web/Web.Shared.Net/WebTestTelemetryInitializer.cs#L36](url)
- Added Unit Test to validate session and user id
- Removed event source method.

abaranch commented:
:shipit:

kartang commented:
Thank you for working on it. Merging this PR.
